One more fix for P3144R2 implementation

hvdijk opened the pull request:
P3144R2 made it ill-formed to delete a pointer to an incomplete class type, and asserted that incomplete class types were the only incomplete types that were possible to pass to the delete operator. This assertion was wrong, and PR 119077 already updated the check to exclude incomplete enum types, but that is still wrong: pointers to arrays of unknown length are another instance of incomplete types that are possible to delete and were not included in P3144R2's ban.

This PR ensures that when we implement the check for incomplete class type, we only check for incomplete class type.

hvdijk committed:
More fixes for P3144R2 implementation

P3144R2 made it ill-formed to delete a pointer to an incomplete class
type, and asserted that incomplete class types were the only incomplete
types that were possible to pass to the delete operator. This assertion
was wrong, and PR 99278 already updated the check to exclude incomplete
enum types, but that is still wrong: pointers to arrays of unknown
length are another instance of incomplete types that are possible to
delete and were not included in PR3144R2's ban. Additionally, the
diagnostic still claimed that the problem was with deleting pointers to
incomplete types, rather than to incomplete class types.

This PR ensures that when we implement the check for incomplete class
type, we only check for incomplete class type, and adjusts the
diagnostics to say 'incomplete struct' or 'incomplete union' to be
accurate without being overly verbose.

hvdijk commented:
Apologies for the repeated review request, not sure why GitHub removed one when I only tried to request an additional reviewer.

cor3ntin reviewed (commented):
This change needs a release note.
Please add an entry to `clang/docs/ReleaseNotes.rst` in the section the most adapted to the change, and referencing any Github issue this change fixes. Thanks!

hvdijk commented:
>  Please add an entry to clang/docs/ReleaseNotes.rst in the section the most adapted to the change, and referencing any Github issue this change fixes. Thanks!

There is no separate GitHub issue for this, but will add a release note.

hvdijk commented on clang/include/clang/Basic/DiagnosticSemaKinds.td:
```diff
@@ -8392,17 +8392,19 @@ def ext_default_init_const : ExtWarn<
   "is a Microsoft extension">,
   InGroup<MicrosoftConstInit>;
 def err_delete_operand : Error<"cannot delete expression of type %0">;
+def err_delete_void_ptr_operand : Error<
+  "cannot delete expression with pointer-to-'void' type %0">;
```
This text is literally copied and pasted from what it says directly below for the pre-C++26 case where it is a warning rather than an error, it's not true that we never say pointer-to-void.

cor3ntin commented on clang/include/clang/Basic/DiagnosticSemaKinds.td:
```diff
@@ -8392,17 +8392,19 @@ def ext_default_init_const : ExtWarn<
 def err_ambiguous_delete_operand : Error<
   "ambiguous conversion of delete expression of type %0 to a pointer">;
 def warn_delete_incomplete : Warning<
-  "deleting pointer to incomplete type %0 is incompatible with C++2c"
+  "deleting pointer to incomplete %select{struct|union}0 %1 is incompatible with C++2c"
```
In C++,  types introduced by `struct`, `class`,  and `union` are all "class types". And for this warning, the distinction does not matter, as they are all treated the same way. So I don't think the added complexity is warranted here

hvdijk replied:
I'm not going to add diagnostics that are technically correct but will confuse users.

hvdijk replied:
> That said, I don't see the value in saying anything other than incomplete type.

My reason for wanting to change the diagnostic is that the diagnostic is telling users that it is an error to delete a pointer to an incomplete type when that is not the case.

That said, I'd be okay with taking the diagnostic changes out of this PR and leaving this as just the fix for the rejects-valid. We can figure out whether we can come up with improved wording in a separate PR.

hvdijk replied:
Done, with this update, the diagnostics are exactly as they were before. I've kept the other updates to the tests, but adjusted the message that is being checked for. If this looks okay we can just merge this. I can open a new issue or PR for the diagnostic later, but since that has proved contentious I'll make sure to write in far more detail about the different options and their advantages and disadvantages.

hvdijk committed:
Update PR.

* Use PointeeRD->getTagKind() instead of RT->isUnionType() for the tag.
* Add release note.

hvdijk commented:
Release note added, and diagnostics updated to use `%select{struct|interface|union|class|enum}` based on `getTagKind()` which is already used for other diagnostics too (look for `|interface|` in the same file).

I do take your point about `pointer-to-'void'` being clumsy wording, despite it already being used, and will have a think about how to rephrase, but I want to avoid the warning suggesting that it is an error to delete a pointer to incomplete types (that was wrong and why I started looking at this in the first place).

hvdijk committed:
Also tweak wording of deleting pointers to void.

hvdijk commented on clang/include/clang/Basic/DiagnosticSemaKinds.td:
```diff
@@ -8392,17 +8392,19 @@ def ext_default_init_const : ExtWarn<
   "ambiguous conversion of delete expression of type %0 to a pointer">;
 def warn_delete_incomplete : Warning<
-  "deleting pointer to incomplete type %0 is incompatible with C++2c"
+  "deleting pointer to incomplete %select{struct|interface|union|class|enum}0 %1 is incompatible with C++2c"
```
Also, there's no reason for a new discussion, I hadn't marked the previous one as resolved. We can continue there.

erichkeane reviewed (changes requested):
Marking as 'requested changes' so this doesn't get inadvertently merged while discussions re: the diagnostics are happening.

hvdijk committed:
Revert diagnostics changes.

hvdijk commented on clang/test/SemaCXX/new-delete.cpp:
```diff
@@ -595,6 +595,10 @@ struct GH99278_2 {
+void GH99278_3(int(*p)[]) {
+  delete p;
+  // expected-warning@-1 {{'delete' applied to a pointer-to-array type 'int (*)[]' treated as 'delete[]'}}
+};
```
What EDG does (thanks Compiler Explorer for having that available) is allow it for arrays of complete class type, and disallow it for arrays of incomplete class type. That seems like a sensible way of going about it, it effectively interprets "If the object being deleted has incomplete class type at the point of deletion, the program is ill-formed" as applying also to subobjects of the object being deleted. Which is not what the standard says, but necessary for what P3144R2 aimed to accomplish, so I'll see if I can implement that.

hvdijk replied:
That turned out to be a trivial one-line change, so I've done it and added tests for it.

hvdijk replied:
I've changed the array tests to use `delete []` so that they test something useful, so that they test something that could be valid even for non-null pointers.

hvdijk replied:
For the codegen crash, I have opened #150359

hvdijk committed:
Add delete pointer to array of class test

Like single class instances, pointers to arrays of complete type are
permitted, pointers to arrays of incomplete type are not. This requires
one small tweak to the diagnostic: we need to check an array's element
for completeness, not the array, in order to allow the S[] test.

hvdijk committed:
Change array deletion tests to use delete [].

hvdijk committed:
One more.

hvdijk committed:
Merge remote-tracking branch 'origin/main' into pr3144r2-fixes
